### PR TITLE
Column schema fix

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #718

### CODE REVIEW

File: src/Schema/ColumnSchemaInterface.php [hunk 1]
```diff
@@ -86,6 +86,8 @@ public function dbType(string|null $value): void;
      * directly in your code.
      *
      * If the value is `null` or an {@see Expression}, there will be no conversion.
+     * In exceptional case (with direct use) you can use array containing the value as the first element and the PDO type as the second
+     * element
```
Comment (samdark) [inline src/Schema/ColumnSchemaInterface.php:90]:
```suggestion
     * element.
```

**Review by samdark — APPROVED** [review]
Please add a line for CHANGELOG.

File: src/Schema/ColumnSchemaInterface.php [hunk 2]
```diff
@@ -86,6 +86,8 @@ public function dbType(string|null $value): void;
      * directly in your code.
      *
      * If the value is `null` or an {@see Expression}, there will be no conversion.
+     * In exceptional case (with direct use) you can use array containing the value as the first element and the PDO type as the second
```
Comment (vjik) [inline src/Schema/ColumnSchemaInterface.php:89]: After this fix for array `['key' => 0, 1 => DataType::STRING]` returns array, but for array `[0, 2]`  returns `Param` object. It's unexpectedly. May be write in phpdoc "do not use directly"?

Protected method `typecast` uses in `phpTypecast()` method. Should we add notice and tests there also?

Comment (darkdef) [reply src/Schema/ColumnSchemaInterface.php:89]: Ok. Later

Comment (darkdef) [reply src/Schema/ColumnSchemaInterface.php]: Comment added.

**Comment by Tigrov** [comment]
The problem with "(with direct use)" when you use array containing the value as the first element and the PDO type as the second you can get unexpected result for array fields, e.g.
```php
$db->createCommand()->createTable('{{%table}}', ['array' => 'json'])->execute(); // or array type instead of json ('integer []')
$db->createCommand()->insert('{{%table}}', ['array' => [1, 2]])->execute();
```
Maybe use a Param object where it is necessary instead of arrays?

**Comment by Tigrov** [comment]
I think there is no real cases where you get array from db with PDO type as the second element. The idea to specify PDO type is to put a php value with correct PDO type into db but not vice versa.
It is not required to create a new Param object when you get values from db.

**Comment by darkdef** [comment]
> It is not required to create a new Param object when you get values from db.
Yes and i was add fix for it behaviour. It's for futuring - non PDO implementations.

Now corrects for your issue?

**Comment by Tigrov** [comment]
> Now corrects for your issue?

```php
$db->createCommand()->createTable('{{%table}}', ['array' => 'json'])->execute(); // or array type instead of json ('integer []')
$db->createCommand()->insert('{{%table}}', ['array' => [1, 2]])->execute();
$result = $db->createCommand('SELECT * FROM {{%table}}')->queryScalar();
// expected $result === '[1, 2]'
```

Will this example work correct?

**Comment by darkdef** [comment]
> 



> Will this example work correct?

In postgresql - yes

```json
string(6) "[1, 2]"
```
But and before fix all been ok (in pgsql)

**Comment by Tigrov** [comment]
Yes, you are right. Thanks.

What about the comment https://github.com/yiisoft/db/issues/718#issuecomment-1576043917

**Comment by darkdef** [comment]
> What about the comment [#718 (comment)](https://github.com/yiisoft/db/issues/718#issuecomment-1576043917)

In other PR. @terabytesoftw work for remove legacy behavior.

File: src/Schema/AbstractColumnSchema.php [hunk 3]
```diff
@@ -245,7 +245,7 @@ public function unsigned(bool $value): void
      *
      * @return mixed The converted value.
      */
-    protected function typecast(mixed $value): mixed
+    protected function typecast(mixed $value, bool $dbCast = true): mixed
```
Comment (vjik) [inline src/Schema/AbstractColumnSchema.php:248]: Need fix phpdoc of this method.

File: src/Schema/AbstractColumnSchema.php [hunk 4]
```diff
@@ -242,10 +242,11 @@ public function unsigned(bool $value): void
      * If the value is null or an {@see Expression}, it won't be converted.
      *
      * @param mixed $value The value to be converted.
+     * @param bool $dbCast Flag for detect direction of type casting (read or write to db)
```
Comment (samdark) [inline src/Schema/AbstractColumnSchema.php:245]:
```suggestion
     * @param bool $dbCast Whether to detect direction of type casting (read or write to db).
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,8 +2,8 @@
 
 ## 1.0.1 under development
 
-- no changes in this release.
+- Bug #718: Сhanging the type casting check condition for value as array (@darkdef)
```
Comment (samdark) [inline CHANGELOG.md:5]:
```suggestion
- Bug #718: Fix passing value => PDO type pair to `AbstractColumnSchema::typecast()` (@darkdef)
```

File: src/Schema/ColumnSchemaInterface.php [hunk 6]
```diff
@@ -86,6 +86,8 @@ public function dbType(string|null $value): void;
      * directly in your code.
      *
      * If the value is `null` or an {@see Expression}, there will be no conversion.
+     * In exceptional case (with direct use) you can use array containing the value as the first element and the PDO
+     * type as the second element.
```
Comment (samdark) [inline src/Schema/ColumnSchemaInterface.php:90]:
```suggestion
     * When using the method directly you can pass an array containing the value
     * as the first element and the PDO type as the second element.
```

File: src/Schema/ColumnSchemaInterface.php [hunk 7]
```diff
@@ -265,6 +267,8 @@ public function phpType(string|null $value): void;
      * Converts the input value according to {@see phpType} after retrieval from the database.
      *
      * If the value is `null` or an {@see Expression}, there is no conversion.
+     * In exceptional case (with direct use) you can use array containing the value as the first element and the PDO
+     * type as the second element.
```
Comment (samdark) [inline src/Schema/ColumnSchemaInterface.php:271]:
```suggestion
     * When using the method directly you can pass an array containing the value
     * as the first element and the PDO type as the second element.
```

**Review by samdark — APPROVED** [review]

**Comment by Tigrov** [comment]
Maybe to move the code 

```php
if (
    $dbCast
    && is_array($value)
    && count($value) === 2
    && isset($value[0], $value[1])
    && in_array($value[1], $this->getDataTypes(), true)
) {
    return new Param((string) $value[0], $value[1]);
}
```

to 

```php
public function dbTypecast(mixed $value): mixed
{
...
}
```

instead of adding a new attribute `bool $dbCast = true`?

**Comment by terabytesoftw** [comment]
> Maybe to move the code
> 
> ```
> if (
>     $dbCast
>     && is_array($value)
>     && count($value) === 2
>     && isset($value[0], $value[1])
>     && in_array($value[1], $this->getDataTypes(), true)
> ) {
>     return new Param((string) $value[0], $value[1]);
> }
> ```
> 
> to
> 
> ```
> public function dbTypecast(mixed $value): mixed
> {
> ...
> }
> ```
> 
> instead of adding a new attribute `bool $dbCast = true`?

In my opinion the best solution is to remove this strange behavior, for me it does not make sense to cast only two values of an array, since in all other scenarios it will fail, it is best to return the values as each dbms returns them, this does not affect at all the operation of all db packages, which i have tested without this code and they all work perfectly.

**Comment by darkdef** [comment]
Not actually - https://github.com/yiisoft/db/pull/722

**Pull request closed without merging**